Compile to records (#67)
reasonml-community/graphql_ppx: README.md

clients include [Reason Apollo Hooks](https://github.com/Astrocoders/reason-apollo-hooks/commits/master)
or [Reason URQL](https://github.com/FormidableLabs/reason-urql). They also
provide a more end-to-end getting started. This documentation will focus on how
to create queries and fragments, and parse responses.

## Defining a Query

You can define a query in your ReasonML file with the following code

```reason
[%graphql {|
  query UserQuery {
    user {
      id
      role
    }
  }
|}];
```

This will create the `UserQuery` module. This module has the following
contents assigned:

### Let bindings

#### Basic

- `query` (`string`), the GraphQL query or mutation
- `parse` (`UserQuery.t_raw => UserQuery.t`), the function to parse the raw
  GraphQL response into ReasonML types.
- `makeVariables` (`(~your, ~arguments, ()) => Js.Json.t`): a
  function that takes labeled arguments to produce the variables that can be
  sent together with the query. This will also validate and type-check the
  variables.
- `definition`: the module contents packaged. This is usually what you provide
  to the client for ergonomics so you don't have to pass multiple arguments per
  query

#### Advanced

> - `serialize` (`t => t_raw`): this is the opposite of parse.

Reviewer: This will be implemented in second PR so we have to remove that from the documentation for now.

  Sometimes you need to convert the ReasonML representation of the response back
  into the raw JSON representation. Usually this is used within the GraphQL
  client for things like updating the internal cache.
- `serializeVariables` (`t_variables => Js.Json.t`): Convert the
  variables (a record) to a Js.Json.t representation as an alternative to the
  labeled function
- `makeInputObject{YourInputObject}` - a labeled function to create
  `YourInputObject`: This is helpful when you have an input object with many
  optional values (works exactly the same as makeVariables)

> - `fromJSON` (`Js.Json.t => t_raw`): With this function you can

Reviewer: Not sure about this function. It's not implemented yet right?

  convert a Js.Json.t response to a `t_raw` response. It is a no-op and just
  casts the type.

### Types

- `t`: the parsed response of the query

> - `t_raw`: the unparsed response. This is basically the exact shape of the raw

Reviewer: Same here.

  response before it is parsed into more ergonomic ReasonML types like `option`
  instead of `Js.Json.t`, variants etc.
- `t_variables`: the variables of the query or mutation

GraphQL objects, variables and input objects are typed as records for `t`,
`t_raw` and `t_variables`. The types are named according to the hierarchy. Each
step in the hierarchy is split using an underscore. So the type of the user
object in the query above is `t_user` if there would be a field that contained
friends of the user it would be called `t_user_friends`.

## Alternative ways of using `%graphql`

When using GraphQL like this:

```reason
module UserQuery = [%graphql {|
  query UserQuery {
    user {
      id
      role
    }
  }
|}];
```

It will have the same effect as the result above. However you can now rename the
query module.

You can also do this:

```reason
module UserQueries = {
  [%graphql {|
    query UserQuery {
      user {
        id
        role
      }
    }
  |}];
};
```

This will create a parent module (the query now will be:
`UserQueries.UserQuery`)

> You can define multiple queries or fragment in within a single GraphQL extension

Reviewer: ```suggestion
You can define multiple operations or fragments within a single GraphQL extension
```

    }
  }
|}]
```

Then `user` will be of the type `Avatar_User.t`.

#### Variables within fragments

Sometimes fragments need to accept variables. Take our previous fragment. If we
would like to pass the pixelRatio as a variable as it might vary per device. We
can do this as follows:

```reason
[%graphql {|
  fragment Avatar_User on User @argumentDefinitions(pixelRatio: {type: "Float!"}) {
    id
    name
    smallAvatar: avatar(pixelRatio: 2, width: 60, height: 60) {
      url
    }
  }

  query UserQuery($pixelRatio: Float!) {
    user {
      id
      role
      ...Avatar_User @arguments(pixelRatio: $pixelRatio)
    }
  }
|}]
```

To be able to typecheck these variables and make sure that the types are correct,
there are no unused variables or variables that are not defined, we introduce
two directives here `argumentDefinitions` and `arguments`, these are taken from
[Relay](https://relay.dev/docs/en/fragment-container#argumentdefinitions). But
they have nothing to do with the relay client (we just re-use this convention).

Note that you cannot rename variables in the `@arguments` directive so the name
of the variable and the name of the key must be the same. This is because

> GraphQL PPX does not manipulaate variable names and just makes use of the fact

Reviewer: ```suggestion
GraphQL PPX does not manipulate variable names and just makes use of the fact
```

that fragments can use variables declared in the query.

There is a compile error raised if you define variables that are unused. If you
(temporarily) want to define unused variables you can prepend the variable name
with an underscore.

#### `bsAs`

An ecape hatch for when you don't want GraphQL PPX to create a record type, you
can supply one yourself. This also makes reusability possible. We recommend
fragments however in most cases as they are easier to work, are safer and don't

> require defining seperate types types.

Reviewer: ```suggestion
require defining separate types.
```

```reason
type t_user = {
  id: string
  role: string
}

[%graphql {|
  query UserQuery {
    user @bsAs(type: "t_user") {
      id
      role
    }
  }
|}]
```

### Custom field decoders

If you've got a custom scalar, or just want to convert e.g. an integer to a
string to properly fit a record type (see above), you can use the `@decoder`
directive to insert a custom function in the decoder:

```reason
module StringHeight = {
  let parse = (height) => string_of_float(height);

>   let serialize = (height) => float_of_string(height);

Reviewer: We don't support it yet.

  type t = string;
}

module HeroQuery = [%graphql {|
{
  hero {
    name
    height @decoder(module: "StringHeight")
    mass
  }
}
|}];
```

In this example, `height` will be converted from a float to a string in the

> result. Using the `modulel` argument, you can specify any decoder module with

Reviewer: ```suggestion
result. Using the `module` argument, you can specify any decoder module with
```
